api/vm/compile.ts

```diff
@@ -1,52 +1,84 @@
 import * as core from "@dataform/core";
+import { dataform } from "@dataform/protos";
 import * as crypto from "crypto";
 import * as fs from "fs";
 import * as os from "os";
 import * as path from "path";
 import { util } from "protobufjs";
 import { NodeVM } from "vm2";
-import { genIndex } from "../gen_index";
-
-export interface ICompileIPCParameters {
-  projectDir: string;
-  schemaSuffixOverride?: string;
-}
+import { createGenIndexConfig } from "./gen_index_config";
 
 export interface ICompileIPCResult {
   path?: string;
   err?: string;
 }
 
-export function compile(projectDir: string, schemaSuffixOverride?: string): Uint8Array {
-  const vm = new NodeVM({
+export function compile(compileConfig: dataform.ICompileConfig): Uint8Array {
+  const vmIndexFileName = path.resolve(path.join(compileConfig.projectDir, "index.js"));
+
+  const indexGeneratorVm = new NodeVM({
+    wrapper: "none",
+    require: {
+      context: "sandbox",
+      root: compileConfig.projectDir,
+      external: true
+    }
+  });
+
+  // TODO: Once all users of @dataform/core are updated to include compiler functions, remove
+  // this exception handling code (and assume existence of genIndex / compiler functions in @dataform/core).
+  const findGenIndex = (): ((base64EncodedConfig: string) => string) => {
+    try {
+      return (
+        indexGeneratorVm.run('return require("@dataform/core").genIndex', vmIndexFileName) ||
+        core.genIndex
+      );
+    } catch (e) {
+      return core.genIndex;
+    }
+  };
+  const findCompiler = (): ((code, path) => string) => {
+    try {
+      return indexGeneratorVm.run(
+        'return require("@dataform/core").compilers.compile',
```

Comment:
Hmm, that is interesting, good to know!

i think we should be careful though, since this is basically defining the API that the core package exposes (potentially for a long time). AFAIUI if I directly `require()` a file, we're forcing the core package to always have a that file/directory structure. alternatively if we use variable names, we should be free to refactor code inside core just as long as we maintain a `compilers.compile` function accessible from the top level of core.

WDYT?

Reply:
good point - I agree!